File: src/rag-module/fingerprint.ts
import { normalizeTemplateText, normalizeText, stableHash } from "./utils";

export function buildExactHash(stem: string, options: string[], answer?: string): string {
  const payload = `${normalizeText(stem)}||${options.map(normalizeText).join("|")}||${normalizeText(answer ?? "")}`;
  return stableHash(payload);
}

export function buildTemplateHash(stem: string): string {
  return stableHash(normalizeTemplateText(stem));
}

File: rag_python/fingerprint.py
from __future__ import annotations

from .utils import normalize_template_text, normalize_text, stable_hash


def build_exact_hash(stem: str, options: list[str], answer: str | None) -> str:
    payload = f"{normalize_text(stem)}||{'|'.join(normalize_text(x) for x in options)}||{normalize_text(answer or '')}"
    return stable_hash(payload)


def build_template_hash(stem: str) -> str:
    return stable_hash(normalize_template_text(stem))
